File: languages/nim/solana-client/src/donation_client.nim
## Nim Solana Donation Client
##
## A high-performance client for interacting with the Solana donation
## smart contract. Combines Python-like syntax with C-like performance.
##
## Example:
##   let client = newDonationClient()
##   let result = client.donate(donorPubkey, 0.5.SOL)
##   echo "Signature: ", result.signature

import std/[options, times, strformat, json, tables, math]
import std/httpclient
import std/asyncdispatch

# =============================================================================
# Constants
# =============================================================================

const
  LamportsPerSol* = 1_000_000_000'u64
  MinDonation* = 1_000_000'u64  # 0.001 SOL
  MaxDonation* = 100_000_000_000'u64  # 100 SOL

  TierBronze* = 1_000_000'u64
  TierSilver* = 100_000_000'u64
  TierGold* = 1_000_000_000'u64
  TierPlatinum* = 10_000_000_000'u64

# =============================================================================
# Types
# =============================================================================

type
  Pubkey* = distinct string
    ## Solana public key (base58 encoded)

  Lamports* = distinct uint64
    ## Amount in lamports (1 SOL = 1,000,000,000 lamports)

  SOL* = distinct float64
    ## Amount in SOL

  DonorTier* = enum
    ## Donor tier classification based on total donations
    Bronze = "bronze"      ## >= 0.001 SOL
    Silver = "silver"      ## >= 0.1 SOL
    Gold = "gold"          ## >= 1 SOL
    Platinum = "platinum"  ## >= 10 SOL

  DonationResult* = object
    ## Result of a donation operation
    signature*: string
    amount*: Lamports
    donor*: Pubkey
    newTier*: DonorTier
    totalDonated*: Lamports

  VaultStats* = object
    ## Vault statistics
    admin*: Pubkey
    totalDonated*: Lamports
    totalWithdrawn*: Lamports
    currentBalance*: Lamports
    donationCount*: uint64
    uniqueDonors*: uint64
    isPaused*: bool
    minDonationAmount*: Lamports
    maxDonationAmount*: Lamports

  DonorInfo* = object
    ## Individual donor information
    donor*: Pubkey
    totalDonated*: Lamports
    donationCount*: uint64
    lastDonationTimestamp*: int64
    tier*: DonorTier

  DonationError* = object of CatchableError
    ## Base exception for donation errors

  DonationTooSmallError* = object of DonationError
  DonationTooLargeError* = object of DonationError
  ContractPausedError* = object of DonationError
  InsufficientFundsError* = object of DonationError
  UnauthorizedError* = object of DonationError

  DonationClient* = ref object
    ## Main client for interacting with donation contract
    rpcUrl: string
    programId: Pubkey
    httpClient: HttpClient

# =============================================================================
# Converters
# =============================================================================

converter toPubkey*(s: string): Pubkey = Pubkey(s)
converter toString*(p: Pubkey): string = string(p)
converter toLamports*(u: uint64): Lamports = Lamports(u)
converter toUint64*(l: Lamports): uint64 = uint64(l)
converter toSOL*(f: float64): SOL = SOL(f)
converter toFloat64*(s: SOL): float64 = float64(s)

# =============================================================================
# Utility Functions
# =============================================================================

proc sol*(amount: float64): Lamports =
  ## Convert SOL to lamports
  ##
  ## Example:
  ##   let lamports = sol(0.5)  # 500_000_000 lamports
  result = (amount * LamportsPerSol.float64).uint64.Lamports

proc toSOL*(lamports: Lamports): float64 =
  ## Convert lamports to SOL
  ##
  ## Example:
  ##   let solAmount = 1_000_000_000.Lamports.toSOL()  # 1.0 SOL
  result = lamports.uint64.float64 / LamportsPerSol.float64

proc calculateTier*(totalDonated: Lamports): DonorTier =
  ## Calculate donor tier based on total donations
  ##
  ## Example:
  ##   let tier = calculateTier(10_000_000_000.Lamports)  # Platinum
  if totalDonated.uint64 >= TierPlatinum:
    Platinum
  elif totalDonated.uint64 >= TierGold:
    Gold
  elif totalDonated.uint64 >= TierSilver:
    Silver
  else:
    Bronze

proc tierEmoji*(tier: DonorTier): string =
  ## Get emoji representation of tier
  ##
  ## Example:
  ##   echo tierEmoji(Platinum)  # "💎"
  case tier
  of Bronze: "🥉"
  of Silver: "🥈"
  of Gold: "🥇"
  of Platinum: "💎"

proc tierName*(tier: DonorTier): string =
  ## Get name of tier
  ##
  ## Example:
  ##   echo tierName(Gold)  # "Gold"
  $tier

proc nextTierThreshold*(currentTier: DonorTier): Option[tuple[tier: DonorTier, threshold: Lamports]] =
  ## Get next tier and its threshold
  ##
  ## Example:
  ##   let next = nextTierThreshold(Bronze)
  ##   if next.isSome:
  ##     echo "Next tier: ", next.get.tier
  case currentTier
  of Bronze:
    some((tier: Silver, threshold: TierSilver.Lamports))
  of Silver:
    some((tier: Gold, threshold: TierGold.Lamports))
  of Gold:
    some((tier: Platinum, threshold: TierPlatinum.Lamports))
  of Platinum:
    none(tuple[tier: DonorTier, threshold: Lamports])

proc lamportsToNextTier*(currentDonated: Lamports, currentTier: DonorTier): Option[Lamports] =
  ## Calculate lamports needed to reach next tier
  ##
  ## Example:
  ##   let needed = lamportsToNextTier(50_000_000.Lamports, Bronze)
  let nextTier = nextTierThreshold(currentTier)
  if nextTier.isNone:
    return none(Lamports)

  let threshold = nextTier.get.threshold
  if currentDonated.uint64 >= threshold.uint64:
    none(Lamports)
  else:
    some((threshold.uint64 - currentDonated.uint64).Lamports)

# =============================================================================
# Validation Functions
# =============================================================================

proc validateDonation*(amount: Lamports, minAmount, maxAmount: Lamports) =
  ## Validate donation amount
  ##
  ## Raises:
  ##   DonationTooSmallError: If amount < minAmount
  ##   DonationTooLargeError: If amount > maxAmount
  if amount.uint64 < minAmount.uint64:
    raise newException(DonationTooSmallError,
      &"Donation of {amount} is below minimum of {minAmount}")

  if amount.uint64 > maxAmount.uint64:
    raise newException(DonationTooLargeError,
      &"Donation of {amount} exceeds maximum of {maxAmount}")

proc checkNotPaused*(stats: VaultStats) =
  ## Check if contract is not paused
  ##
  ## Raises:
  ##   ContractPausedError: If contract is paused
  if stats.isPaused:
    raise newException(ContractPausedError, "Contract is currently paused")

proc checkSufficientFunds*(required, available: Lamports) =
  ## Check if sufficient funds are available
  ##
  ## Raises:
  ##   InsufficientFundsError: If available < required
  if available.uint64 < required.uint64:
    raise newException(InsufficientFundsError,
      &"Insufficient funds: need {required}, have {available}")

proc checkAuthorization*(caller, expected: Pubkey) =
  ## Check if caller is authorized
  ##
  ## Raises:
  ##   UnauthorizedError: If caller != expected
  if caller.string != expected.string:
    raise newException(UnauthorizedError,
      &"Unauthorized: expected {expected}, got {caller}")

# =============================================================================
# Client Implementation
# =============================================================================

proc newDonationClient*(
  rpcUrl: string = "https://api.devnet.solana.com",
  programId: string = "DoNaT1on111111111111111111111111111111111111"
): DonationClient =
  ## Create a new donation client
  ##
  ## Example:
  ##   let client = newDonationClient()
  ##   let client2 = newDonationClient("https://api.mainnet-beta.solana.com")
  result = DonationClient(
    rpcUrl: rpcUrl,
    programId: programId.Pubkey,
    httpClient: newHttpClient()
  )

proc donate*(
  client: DonationClient,
  donor: Pubkey,
  amount: SOL
): DonationResult =
  ## Make a donation to the vault
  ##
  ## Args:
  ##   donor: Donor's public key
  ##   amount: Amount in SOL
  ##
  ## Returns:
  ##   DonationResult with signature and updated info
  ##
  ## Raises:
  ##   DonationTooSmallError, DonationTooLargeError, ContractPausedError
  ##
  ## Example:
  ##   let result = client.donate(donorPubkey, 0.5.SOL)
  ##   echo "Signature: ", result.signature

  let lamports = sol(amount.float64)

  # Validate donation amount
  validateDonation(lamports, MinDonation.Lamports, MaxDonation.Lamports)

  # Check if contract is paused
  let stats = client.getVaultStats()
  checkNotPaused(stats)

  # Send transaction (placeholder)
  let signature = "signature_" & $hash(donor.string)

  # Get updated donor info (placeholder)
  let donorInfo = client.getDonorInfo(donor)

  result = DonationResult(
    signature: signature,
    amount: lamports,
    donor: donor,
    newTier: donorInfo.tier,
    totalDonated: donorInfo.totalDonated
  )

proc getVaultStats*(client: DonationClient): VaultStats =
  ## Get vault statistics
  ##
  ## Returns:
  ##   VaultStats with current vault information
  ##
  ## Example:
  ##   let stats = client.getVaultStats()
  ##   echo "Total donated: ", stats.totalDonated.toSOL(), " SOL"

  # Placeholder implementation - would call RPC endpoint
  result = VaultStats(
    admin: "Admin123...".Pubkey,
    totalDonated: 1_000_000_000'u64.Lamports,
    totalWithdrawn: 0'u64.Lamports,
    currentBalance: 1_000_000_000'u64.Lamports,
    donationCount: 10'u64,
    uniqueDonors: 5'u64,
    isPaused: false,
    minDonationAmount: MinDonation.Lamports,
    maxDonationAmount: MaxDonation.Lamports
  )

proc getDonorInfo*(client: DonationClient, donor: Pubkey): DonorInfo =
  ## Get donor information
  ##
  ## Args:
  ##   donor: Donor's public key
  ##
  ## Returns:
  ##   DonorInfo with donor statistics and tier
  ##
  ## Example:
  ##   let info = client.getDonorInfo(donorPubkey)
  ##   echo "Tier: ", info.tier.tierEmoji()

  # Placeholder implementation - would call RPC endpoint
  let totalDonated = 500_000_000'u64.Lamports
  result = DonorInfo(
    donor: donor,
    totalDonated: totalDonated,
    donationCount: 3'u64,
    lastDonationTimestamp: getTime().toUnix(),
    tier: calculateTier(totalDonated)
  )

proc withdraw*(
  client: DonationClient,
  admin: Pubkey,
  amount: SOL
): string =
  ## Withdraw funds from the vault (admin only)
  ##
  ## Args:
  ##   admin: Admin's public key
  ##   amount: Amount to withdraw in SOL
  ##
  ## Returns:
  ##   Transaction signature
  ##
  ## Raises:
  ##   UnauthorizedError, InsufficientFundsError
  ##
  ## Example:
  ##   let signature = client.withdraw(adminPubkey, 1.0.SOL)

  let lamports = sol(amount.float64)

  # Get vault stats
  let stats = client.getVaultStats()

  # Check authorization
  checkAuthorization(admin, stats.admin)

  # Check sufficient funds
  checkSufficientFunds(lamports, stats.currentBalance)

  # Send transaction (placeholder)
  result = "withdraw_signature_" & $hash(admin.string)

# =============================================================================
# Formatting and Display
# =============================================================================

proc `$`*(tier: DonorTier): string =
  ## String representation of tier
  tierName(tier) & " " & tierEmoji(tier)

proc `$`*(lamports: Lamports): string =
  ## String representation of lamports
  &"{lamports.toSOL():.9f} SOL ({lamports.uint64} lamports)"

proc `$`*(result: DonationResult): string =
  ## String representation of donation result
  &"""
Donation Result:
  Signature: {result.signature}
  Amount: {result.amount}
  Donor: {result.donor}
  New Tier: {result.newTier}
  Total Donated: {result.totalDonated}
"""

proc `$`*(stats: VaultStats): string =
  ## String representation of vault stats
  &"""
Vault Statistics:
  Admin: {stats.admin}
  Total Donated: {stats.totalDonated}
  Total Withdrawn: {stats.totalWithdrawn}
  Current Balance: {stats.currentBalance}
  Donations: {stats.donationCount}
  Unique Donors: {stats.uniqueDonors}
  Paused: {stats.isPaused}
"""

# =============================================================================
# Example Usage
# =============================================================================

when isMainModule:
  echo "=== Nim Solana Donation Client ==="

  # Create client
  let client = newDonationClient()
  echo "✓ Client created"

  # Example donor
  let donor = "DonorPubkey123...".Pubkey

  # Make donation
  try:
    let result = client.donate(donor, 0.5.SOL)
    echo "\n✓ Donation successful:"
    echo result
  except DonationError as e:
    echo "✗ Donation failed: ", e.msg

  # Get vault stats
  let stats = client.getVaultStats()
  echo "\n✓ Vault stats:"
  echo stats

  # Get donor info
  let info = client.getDonorInfo(donor)
  echo "\n✓ Donor info:"
  echo &"  Tier: {info.tier}"
  echo &"  Total: {info.totalDonated}"
  echo &"  Count: {info.donationCount}"

  # Check next tier
  let nextTier = lamportsToNextTier(info.totalDonated, info.tier)
  if nextTier.isSome:
    echo &"  Need {nextTier.get.toSOL():.4f} SOL for next tier"
  else:
    echo "  Already at max tier! 💎"

File: languages/zig/solana-client/src/donation_client.zig
// Zig Solana Donation Client
//
// A high-performance, memory-safe client for interacting with the
// Solana donation smart contract. Features explicit memory management
// and compile-time safety guarantees.
//
// Example:
//   var client = try DonationClient.init(allocator, .{});
//   defer client.deinit();
//   const result = try client.donate(donor, 0.5);

const std = @import("std");
const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;

// =============================================================================
// Constants
// =============================================================================

pub const LAMPORTS_PER_SOL: u64 = 1_000_000_000;
pub const MIN_DONATION: u64 = 1_000_000; // 0.001 SOL
pub const MAX_DONATION: u64 = 100_000_000_000; // 100 SOL

pub const TIER_BRONZE: u64 = 1_000_000;
pub const TIER_SILVER: u64 = 100_000_000;
pub const TIER_GOLD: u64 = 1_000_000_000;
pub const TIER_PLATINUM: u64 = 10_000_000_000;

// =============================================================================
// Types
// =============================================================================

/// Donor tier classification based on total donations
pub const DonorTier = enum {
    bronze, // >= 0.001 SOL
    silver, // >= 0.1 SOL
    gold, // >= 1 SOL
    platinum, // >= 10 SOL

    /// Calculate tier from lamports amount
    pub fn fromLamports(lamports: u64) DonorTier {
        if (lamports >= TIER_PLATINUM) return .platinum;
        if (lamports >= TIER_GOLD) return .gold;
        if (lamports >= TIER_SILVER) return .silver;
        return .bronze;
    }

    /// Get emoji representation of tier
    pub fn emoji(self: DonorTier) []const u8 {
        return switch (self) {
            .bronze => "🥉",
            .silver => "🥈",
            .gold => "🥇",
            .platinum => "💎",
        };
    }

    /// Get name of tier
    pub fn name(self: DonorTier) []const u8 {
        return switch (self) {
            .bronze => "Bronze",
            .silver => "Silver",
            .gold => "Gold",
            .platinum => "Platinum",
        };
    }

    /// Get next tier and its threshold
    pub fn nextTier(self: DonorTier) ?struct { tier: DonorTier, threshold: u64 } {
        return switch (self) {
            .bronze => .{ .tier = .silver, .threshold = TIER_SILVER },
            .silver => .{ .tier = .gold, .threshold = TIER_GOLD },
            .gold => .{ .tier = .platinum, .threshold = TIER_PLATINUM },
            .platinum => null,
        };
    }
};

/// Amount in lamports with conversion utilities
pub const Lamports = struct {
    value: u64,

    /// Create from SOL amount
    pub fn fromSol(sol: f64) Lamports {
        return .{ .value = @intFromFloat(sol * @as(f64, @floatFromInt(LAMPORTS_PER_SOL))) };
    }

    /// Convert to SOL
    pub fn toSol(self: Lamports) f64 {
        return @as(f64, @floatFromInt(self.value)) / @as(f64, @floatFromInt(LAMPORTS_PER_SOL));
    }

    /// Format for printing
    pub fn format(
        self: Lamports,
        comptime fmt: []const u8,
        options: std.fmt.FormatOptions,
        writer: anytype,
    ) !void {
        _ = fmt;
        _ = options;
        try writer.print("{d} lamports ({d:.9} SOL)", .{ self.value, self.toSol() });
    }
};

/// Result of a donation operation
pub const DonationResult = struct {
    signature: []const u8,
    amount: Lamports,
    donor: []const u8,
    new_tier: DonorTier,
    total_donated: Lamports,

    pub fn deinit(self: *DonationResult, allocator: Allocator) void {
        allocator.free(self.signature);
        allocator.free(self.donor);
    }
};

/// Vault statistics
pub const VaultStats = struct {
    admin: []const u8,
    total_donated: Lamports,
    total_withdrawn: Lamports,
    current_balance: Lamports,
    donation_count: u64,
    unique_donors: u64,
    is_paused: bool,
    min_donation_amount: Lamports,
    max_donation_amount: Lamports,

    pub fn deinit(self: *VaultStats, allocator: Allocator) void {
        allocator.free(self.admin);
    }
};

/// Donor information
pub const DonorInfo = struct {
    donor: []const u8,
    total_donated: Lamports,
    donation_count: u64,
    last_donation_timestamp: i64,
    tier: DonorTier,

    /// Calculate lamports needed to reach next tier
    pub fn lamportsToNextTier(self: DonorInfo) ?Lamports {
        const next = self.tier.nextTier() orelse return null;
        const threshold = Lamports{ .value = next.threshold };

        if (self.total_donated.value >= threshold.value) return null;

        return Lamports{ .value = threshold.value - self.total_donated.value };
    }

    pub fn deinit(self: *DonorInfo, allocator: Allocator) void {
        allocator.free(self.donor);
    }
};

// =============================================================================
// Errors
// =============================================================================

pub const DonationError = error{
    DonationTooSmall,
    DonationTooLarge,
    ContractPaused,
    InsufficientFunds,
    Unauthorized,
    NetworkError,
    ParseError,
    OutOfMemory,
};

// =============================================================================
// Client Configuration
// =============================================================================

pub const ClientConfig = struct {
    rpc_url: []const u8 = "https://api.devnet.solana.com",
    program_id: []const u8 = "DoNaT1on111111111111111111111111111111111111",
};

// =============================================================================
// Main Client
// =============================================================================

pub const DonationClient = struct {
    allocator: Allocator,
    rpc_url: []const u8,
    program_id: []const u8,

    /// Initialize a new donation client
    ///
    /// Example:
    ///   var client = try DonationClient.init(allocator, .{});
    ///   defer client.deinit();
    pub fn init(allocator: Allocator, config: ClientConfig) !DonationClient {
        return DonationClient{
            .allocator = allocator,
            .rpc_url = config.rpc_url,
            .program_id = config.program_id,
        };
    }

    /// Clean up resources
    pub fn deinit(self: *DonationClient) void {
        _ = self;
        // Cleanup if needed
    }

    /// Make a donation to the vault
    ///
    /// Args:
    ///   donor: Donor's public key
    ///   amount_sol: Amount in SOL
    ///
    /// Returns:
    ///   DonationResult with signature and updated info
    ///
    /// Example:
    ///   const result = try client.donate(donor, 0.5);
    ///   defer result.deinit(allocator);
    pub fn donate(
        self: *DonationClient,
        donor: []const u8,
        amount_sol: f64,
    ) !DonationResult {
        const amount = Lamports.fromSol(amount_sol);

        // Validate donation amount
        try self.validateDonation(amount);

        // Check if contract is paused
        var stats = try self.getVaultStats();
        defer stats.deinit(self.allocator);

        if (stats.is_paused) return error.ContractPaused;

        // Send transaction (placeholder)
        const signature = try std.fmt.allocPrint(
            self.allocator,
            "signature_{d}",
            .{std.crypto.random.int(u64)},
        );

        // Get updated donor info
        var donor_info = try self.getDonorInfo(donor);
        defer donor_info.deinit(self.allocator);

        const donor_copy = try self.allocator.dupe(u8, donor);

        return DonationResult{
            .signature = signature,
            .amount = amount,
            .donor = donor_copy,
            .new_tier = donor_info.tier,
            .total_donated = donor_info.total_donated,
        };
    }

    /// Get vault statistics
    ///
    /// Returns:
    ///   VaultStats with current vault information
    ///
    /// Example:
    ///   const stats = try client.getVaultStats();
    ///   defer stats.deinit(allocator);
    pub fn getVaultStats(self: *DonationClient) !VaultStats {
        // Placeholder implementation - would call RPC endpoint
        const admin = try self.allocator.dupe(u8, "Admin123...");

        return VaultStats{
            .admin = admin,
            .total_donated = .{ .value = 1_000_000_000 },
            .total_withdrawn = .{ .value = 0 },
            .current_balance = .{ .value = 1_000_000_000 },
            .donation_count = 10,
            .unique_donors = 5,
            .is_paused = false,
            .min_donation_amount = .{ .value = MIN_DONATION },
            .max_donation_amount = .{ .value = MAX_DONATION },
        };
    }

    /// Get donor information
    ///
    /// Args:
    ///   donor: Donor's public key
    ///
    /// Returns:
    ///   DonorInfo with donor statistics and tier
    ///
    /// Example:
    ///   const info = try client.getDonorInfo(donor);
    ///   defer info.deinit(allocator);
    pub fn getDonorInfo(self: *DonationClient, donor: []const u8) !DonorInfo {
        // Placeholder implementation - would call RPC endpoint
        const donor_copy = try self.allocator.dupe(u8, donor);
        const total_donated = Lamports{ .value = 500_000_000 };

        return DonorInfo{
            .donor = donor_copy,
            .total_donated = total_donated,
            .donation_count = 3,
            .last_donation_timestamp = std.time.timestamp(),
            .tier = DonorTier.fromLamports(total_donated.value),
        };
    }

    /// Withdraw funds from the vault (admin only)
    ///
    /// Args:
    ///   admin: Admin's public key
    ///   amount_sol: Amount to withdraw in SOL
    ///
    /// Returns:
    ///   Transaction signature
    ///
    /// Example:
    ///   const sig = try client.withdraw(admin, 1.0);
    ///   defer allocator.free(sig);
    pub fn withdraw(
        self: *DonationClient,
        admin: []const u8,
        amount_sol: f64,
    ) ![]const u8 {
        const amount = Lamports.fromSol(amount_sol);

        // Get vault stats
        var stats = try self.getVaultStats();
        defer stats.deinit(self.allocator);

        // Check authorization
        if (!std.mem.eql(u8, admin, stats.admin)) {
            return error.Unauthorized;
        }

        // Check sufficient funds
        if (amount.value > stats.current_balance.value) {
            return error.InsufficientFunds;
        }

        // Send transaction (placeholder)
        return try std.fmt.allocPrint(
            self.allocator,
            "withdraw_signature_{d}",
            .{std.crypto.random.int(u64)},
        );
    }

    // =========================================================================
    // Private Methods
    // =========================================================================

    fn validateDonation(self: *DonationClient, amount: Lamports) !void {
        _ = self;
        if (amount.value < MIN_DONATION) return error.DonationTooSmall;
        if (amount.value > MAX_DONATION) return error.DonationTooLarge;
    }
};

// =============================================================================
// Utility Functions
// =============================================================================

/// Convert SOL to lamports
pub fn sol(amount: f64) Lamports {
    return Lamports.fromSol(amount);
}

/// Calculate donation percentage
pub fn donationPercentage(donor_amount: Lamports, total_amount: Lamports) f64 {
    if (total_amount.value == 0) return 0.0;
    return (@as(f64, @floatFromInt(donor_amount.value)) / @as(f64, @floatFromInt(total_amount.value))) * 100.0;
}

/// Format time ago
pub fn formatTimeAgo(allocator: Allocator, timestamp: i64) ![]const u8 {
    const now = std.time.timestamp();
    const diff = now - timestamp;

    const days = @divTrunc(diff, 86400);
    const hours = @divTrunc(@mod(diff, 86400), 3600);
    const minutes = @divTrunc(@mod(diff, 3600), 60);

    if (days > 0) {
        return try std.fmt.allocPrint(allocator, "{d} days ago", .{days});
    } else if (hours > 0) {
        return try std.fmt.allocPrint(allocator, "{d} hours ago", .{hours});
    } else if (minutes > 0) {
        return try std.fmt.allocPrint(allocator, "{d} minutes ago", .{minutes});
    } else {
        return try allocator.dupe(u8, "just now");
    }
}

// =============================================================================
// Example Usage
// =============================================================================

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("=== Zig Solana Donation Client ===\n\n", .{});

    // Create client
    var client = try DonationClient.init(allocator, .{});
    defer client.deinit();
    std.debug.print("✓ Client created\n", .{});

    // Example donor
    const donor = "DonorPubkey123...";

    // Make donation
    std.debug.print("\n✓ Making donation...\n", .{});
    var result = try client.donate(donor, 0.5);
    defer result.deinit(allocator);

    std.debug.print("  Signature: {s}\n", .{result.signature});
    std.debug.print("  Amount: {}\n", .{result.amount});
    std.debug.print("  Tier: {s} {s}\n", .{ result.new_tier.name(), result.new_tier.emoji() });

    // Get vault stats
    std.debug.print("\n✓ Vault statistics:\n", .{});
    var stats = try client.getVaultStats();
    defer stats.deinit(allocator);

    std.debug.print("  Total donated: {}\n", .{stats.total_donated});
    std.debug.print("  Current balance: {}\n", .{stats.current_balance});
    std.debug.print("  Donations: {d}\n", .{stats.donation_count});
    std.debug.print("  Unique donors: {d}\n", .{stats.unique_donors});

    // Get donor info
    std.debug.print("\n✓ Donor information:\n", .{});
    var info = try client.getDonorInfo(donor);
    defer info.deinit(allocator);

    std.debug.print("  Total: {}\n", .{info.total_donated});
    std.debug.print("  Count: {d}\n", .{info.donation_count});
    std.debug.print("  Tier: {s} {s}\n", .{ info.tier.name(), info.tier.emoji() });

    // Check next tier
    if (info.lamportsToNextTier()) |needed| {
        std.debug.print("  Need {d:.4} SOL for next tier\n", .{needed.toSol()});
    } else {
        std.debug.print("  Max tier reached! 💎\n", .{});
    }
}
